examples/dataset_example/src/dataset_example.cc

```diff
@@ -0,0 +1,239 @@
+// Copyright 2024 TGS
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//    http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file Provides an example of the basic usage of an MDIO Dataset.
+ * Goals:
+ * 1. Starting with an examlpe dataset, explore the variables and metadata.
+ * 2. Understand the in memory SharedArray and on disk representations of the
+ * data.
+ * 3. Read and write with a slicing operator.
+ * 4. Understand Future and Result operators.
+ * library.
+ */
+
+#include "dataset_example.h"
+
+#include <filesystem>
+
+#include "mdio/dataset.h"
+
+using Index = mdio::Index;
+using namespace mdio::dtypes;
+
+/**
+ * @brief This provides a basic MDIO example, covering its API usage and key
+ * concepts.
+ */
+absl::Status Run() {
+  /// Datasets create a number of zarr "files" which are directories
+  /// with json metadata.
+  std::filesystem::path temp_dir = std::filesystem::temp_directory_path();
+
+  std::string dataset_path = temp_dir / "zarrs/seismic_3d/";
+
+  auto json_spec = GetToyExample();
+
+  /// A dataset is used to represent a collection of data "on disk".
+  /// The MDIO json schema for this can be found here:
+  ///    mdio/dataset_schema.h
+  auto dataset_future = mdio::Dataset::from_json(json_spec, dataset_path,
+                                                 mdio::constants::kCreateClean);
+
+  /// MDIO Datasets are compatible with Python's Xarray format:
+  /// https://docs.xarray.dev/en/stable/
+  /// The can be created or loaded using Xarray.
+
+  /// Creating a new dataset will write json data to disk. There maybe some
+  /// latency in doing these write, particularly to object stores. The return
+  /// type of this operation is a Tensorstore Future. A Future allows this I/O
+  /// to happen without blocking the main thread.
+  auto dataset_result = dataset_future.result();
+  /// Here result() Blocks so we have result for the next step.
+
+  /// The MDIO library is not intended to throw exceptions, instead errors are
+  /// handled by returning a "Result" object; the result can be tested for
+  /// "ok()", there is also a convenient Tensorstore macro for doing assigment
+  /// while handling errors.
+  MDIO_ASSIGN_OR_RETURN(auto dataset, dataset_result)
+
+  /// The dataset represent data on disk, the object holds only minimal state.
+  /// Its memory footprint is small.
+  std::cout << dataset << "\n" << std::endl;
+
+  /// Variable type is discoverable at runtime, initially the variable will have
```

Comment:
I might consider using the `at(std::string name)` method over `get<void>(std::string name);` here. It provides a lighter weight method to query the Variable properties' as demonstrated.